test/e2e/functional_test.go

```diff
@@ -183,6 +184,81 @@ func (s *FunctionalSuite) TestFlatmapUDF() {
 	w.Expect().VertexPodLogContains("out", "hello", PodLogCheckOptionWithCount(3))
 }
 
+func (s *FunctionalSuite) TestWatermarkEnabled() {
+	w := s.Given().Pipeline("@testdata/watermark.yaml").
+		When().
+		CreatePipelineAndWait()
+	defer w.DeletePipelineAndWait()
+
+	pipelineName := "simple-pipeline-watermark"
+	// TODO: Any way to extract the list from suite
+	vertexList := []string{"input", "cat1", "cat2", "cat3", "output1", "output2"}
+
+	w.Expect().
+		VertexPodsRunning().DaemonPodsRunning().
+		VertexPodLogContains("input", LogSourceVertexStarted).
+		VertexPodLogContains("cat1", LogUDFVertexStarted, PodLogCheckOptionWithContainer("main")).
+		VertexPodLogContains("cat2", LogUDFVertexStarted, PodLogCheckOptionWithContainer("main")).
+		VertexPodLogContains("cat3", LogUDFVertexStarted, PodLogCheckOptionWithContainer("main")).
+		VertexPodLogContains("output1", LogSinkVertexStarted).
+		VertexPodLogContains("output2", LogSinkVertexStarted).
+		DaemonPodLogContains(pipelineName, LogDaemonStarted).
+		VertexPodLogContains("output1", `"Data":".*","Createdts":.*`).
+		VertexPodLogContains("output2", `"Data":".*","Createdts":.*`)
+
+	defer w.DaemonPodPortForward(pipelineName, 1234, dfv1.DaemonServicePort).
+		TerminateAllPodPortForwards()
+
+	// Test Daemon service with gRPC
+	client, err := daemonclient.NewDaemonServiceClient("localhost:1234")
+	assert.NoError(s.T(), err)
+	buffers, err := client.ListPipelineBuffers(context.Background(), pipelineName)
+	assert.NoError(s.T(), err)
+	assert.Equal(s.T(), 5, len(buffers))
+	bufferInfo, err := client.GetPipelineBuffer(context.Background(), pipelineName, dfv1.GenerateEdgeBufferName(Namespace, pipelineName, "input", "cat1"))
+	assert.NoError(s.T(), err)
+	assert.Equal(s.T(), "input", *bufferInfo.FromVertex)
+
+	var wg sync.WaitGroup
+	for _, vertex := range vertexList {
+		wg.Add(1)
+		go func(vertex string) {
+			defer wg.Done()
+			// timeout for checking watermark progression
+			ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+			defer cancel()
+			go func() {
+				<-ctx.Done()
+				if ctx.Err() == context.DeadlineExceeded {
+					assert.Failf(s.T(), "isWatermarkProgressing timed out", "")
+				}
+			}()
+			assert.Truef(s.T(), isWatermarkProgressing(ctx, client, pipelineName, vertex, 3), "isWatermarkProgressing")
```

Comment:
If `isWatermarkProgressing` takes more than 1 minute, even though the spawned goroutine has asserted failure, but this guy will not end, which means the `wg` will not end, is that right?

Normally, we do the reverse way like below:

```golang
ctx, cancel := xxxx
defer cancel()
done := make(chan struct{})
go func() {
  isWatermarkProgressing...
  close(done)
}()
select {
  case <- ctx.Done():
    assert.Fail.....
  case <- done:
}
```